File: Neetcode/Graphs/surroundedregion.cpp
#include <vector>
using namespace std;

class Solution {
 public:
  void dfs(vector<vector<char>> &board, int i, int j) {
    if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size() ||
        board[i][j] != 'O') {
      return;
    }
    board[i][j] = 'o';

    dfs(board, i + 1, j);
    dfs(board, i - 1, j);
    dfs(board, i, j - 1);
    dfs(board, i, j + 1);
  }

  void solve(vector<vector<char>> &board) {
    int n = board.size();
    int m = board[0].size();

    for (int i = 0; i < n; i++) {
      dfs(board, i, 0);
      dfs(board, i, m - 1);
    }

    for (int j = 1; j < m - 1; j++) {
      dfs(board, 0, j);
      dfs(board, n - 1, j);
    }

    for (int i = 0; i < n; i++) {
      for (int j = 0; j < m; j++) {
        if (board[i][j] == 'o') {
          board[i][j] = 'O';
        } else {
          board[i][j] = 'X';
        }
      }
    }
  }
};

File: Neetcode/Graphs/surroundedregion.py
# https://neetcode.io/problems/surrounded-regions
# https://leetcode.com/problems/surrounded-regions/

from typing import List


class Solution:
    def solve(self, board: List[List[str]]) -> None:
        # check if O island ever touches border
        HEIGHT = len(board)
        WIDTH = len(board[0])

        def dfs(i, j):
            if i < 0 or i >= HEIGHT or j < 0 or j >= WIDTH or board[i][j] != "O":
                return
            board[i][j] = "o"
            for x, y in [[1, 0], [-1, 0], [0, 1], [0, -1]]:
                dfs(i + x, y + j)

        # Look at all 4 edges
        for i in range(HEIGHT):
            if board[i][0] == "O":
                dfs(i, 0)
            if board[i][WIDTH - 1] == "O":
                dfs(i, WIDTH - 1)

        for j in range(1, WIDTH - 1):
            if board[0][j] == "O":
                dfs(0, j)
            if board[HEIGHT - 1][j] == "O":
                dfs(HEIGHT - 1, j)

        for i in range(HEIGHT):
            for j in range(WIDTH):
                if board[i][j] == "o":
                    board[i][j] = "O"
                else:
                    board[i][j] = "X"
